### PR TITLE
Add option to use custom fs

### PR DESCRIPTION
**Pull request by younes200**
Useful with `atom/electron` which use a [custom fs](https://github.com/atom/electron/blob/master/docs/tutorial/application-packaging.md) and you want to  pass the original `fs` as option to `rimraf`

### CODE REVIEW

**Comment by yoshuawuyts** [comment]
What are the cases in `atom/electron` when this is useful?

**Comment by younes200** [comment]
`atom/electron` use [patched](https://github.com/atom/electron/blob/master/docs/tutorial/application-packaging.md) version of `fs` to deal with `asar` archive. 
In electron's context `rimraf` will fail with `ENOTEMPTY` if the folder contains a `asar` file.

 [mkdirp](https://github.com/substack/node-mkdirp) provides the same 'options.fs' to deal with this case so it will be nice to have it in `rimraf` too.

**Comment by yoshuawuyts** [comment]
Sounds like a legit use case, would be pretty useful to have. I think the only thing that's missing from this patch is tests + documentation, but the implementation looks good to me.

**Comment by younes200** [comment]
Done, I add tests for both async/sync using mock-fs.

File: test/basic.js [hunk 1]
```diff
@@ -91,16 +95,97 @@ t.test('no glob', function (t) {
...
+
+
+
+
 t.test('verify that cleanup happened', function (t) {
   t.throws(fs.statSync.bind(fs, __dirname + '/../target'))
   t.throws(fs.statSync.bind(fs, __dirname + '/target'))
   t.end()
-})
+})
```
Comment (yoshuawuyts) [inline test/basic.js]: newline is missing

**Comment by yoshuawuyts** [comment]
apart from the newline gone missing everything lgtm!

**Comment by isaacs** [comment]
You can already do this with the current set of options.  Just override the functions that you want to override.

```javascript
var mockfs = require('some-kind-of-mock-fs')
rimraf('some-path', mockfs, callback)
```

Or, if you just want to override specific methods:

```javascript
rimraf('some-path', {
  stat: myMockStat,
  statSync: myMockStatSync,
  readdir: myMockReaddir,
  // etc
}, callback)
```

Can you explain what use-case this PR allows that isn't already supported?

**Comment by jprichardson** [comment]
> You can already do this with the current set of options. Just override the functions that you want to override.

I suspect this snuck in at some point then: https://github.com/isaacs/rimraf/blob/ef032b689231da56a55cdd62502dbbfb24c36341/rimraf.js#L64

**Comment by younes200** [comment]
> You can already do this with the current set of options. Just override the functions that you want to override.

@isaacs Actually, no you can't, that the point of my PR, check the function signature :
https://github.com/isaacs/rimraf/blob/master/rimraf.js#L43

**Comment by isaacs** [comment]
@younes200 Yes, you definitely can.  Here's an example:

```javascript
var rimraf = require('rimraf')
var fs = require('fs')
// Pass in the default fs functions for all but the ones I want to intercept
// This could also be a whole-hog virtual fs thing
var myFs = Object.create(fs, {
  unlink: { value: function (file, cb) {
    console.error('intercept unlinking %s', file)
    return fs.unlink(file, cb)
  }},
  unlinkSync: { value: function (file) {
    console.error('intercept unlinking sync %s', file)
    return fs.unlinkSync(file)
  }},
  rmdir: { value: function (dir, cb) {
    console.error('intercept rmdir %s', dir)
    return fs.rmdir(dir, cb)
  }},
  rmdirSync: { value: function (dir) {
    console.error('intercept rmdirSync %s', dir)
    return fs.rmdirSync(dir)
  }},
  lstat: { value: function (path, cb) {
    console.error('intercept lstat %s', path)
    return fs.lstat(path, cb)
  }},
  lstatSync: { value: function (path) {
    console.error('intercept lstat %s', path)
    return fs.lstatSync(path)
  }}
})

var mkdirp = require('mkdirp')

function create () {
  ;['a', 'b', 'c'].forEach(function (i) {
    ;['x', 'y', 'z'].forEach(function (j) {
      mkdirp.sync(i + '/' + j)
      fs.writeFileSync(i + '/' + j + '/some-file.txt', 'test\n')
    })
  })
}


create()
rimraf('a', myFs, function (er) {
  if (er) {
    throw er
  }
  console.error('finished with intercepted rimraf')
  create()
  rimraf.sync('a', myFs)
  console.error('finished with intercepted rimraf.sync')
  create()
  rimraf.sync('a')
  console.error('finished with non-intercepted rimraf.sync')
})
```

Output:

```
intercept lstat a
intercept rmdir a
intercept lstat a/x
intercept lstat a/y
intercept lstat a/z
intercept rmdir a/x
intercept rmdir a/y
intercept rmdir a/z
intercept lstat a/z/some-file.txt
intercept lstat a/y/some-file.txt
intercept lstat a/x/some-file.txt
not actually unlinking a/z/some-file.txt
not actually unlinking a/y/some-file.txt
not actually unlinking a/x/some-file.txt
intercept rmdir a/z
intercept rmdir a/y
intercept rmdir a/x
intercept rmdir a
finished with intercepted rimraf
intercept lstat a
intercept rmdirSync a
intercept lstat a/x
intercept rmdirSync a/x
intercept lstat a/x/some-file.txt
intercept unlinking sync a/x/some-file.txt
intercept rmdirSync a/x
intercept lstat a/y
intercept rmdirSync a/y
intercept lstat a/y/some-file.txt
intercept unlinking sync a/y/some-file.txt
intercept rmdirSync a/y
intercept lstat a/z
intercept rmdirSync a/z
intercept lstat a/z/some-file.txt
intercept unlinking sync a/z/some-file.txt
intercept rmdirSync a/z
intercept rmdirSync a
finished with intercepted rimraf.sync
finished with non-intercepted rimraf.sync
```

**Pull request closed without merging**